webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/MergeNodeProperties.scala

```diff
@@ -0,0 +1,389 @@
+   * last children in tail
+   */
+  def builtTrees(targets: List[FullRuleTargetInfo]): PureResult[List[InvertedTree[GroupProp]]] = {
+    /*
+     * The recusrive function: process group in `remains` and add it to `done`.
+     * Each step can fail if we miss some group reference.
+     */
+    def recurseTrees(allTargets: Map[NodeGroupId, GroupProp], remain: List[GroupProp], done: List[InvertedTree[GroupProp]]): PureResult[List[InvertedTree[GroupProp]]] = {
+      remain match {
+        case Nil => Right(done)
+        case h::tail =>
+          buildOneTree(h, allTargets).flatMap(tree =>
+            // remove groups which were processed in that hierarchy
+            recurseTrees(allTargets, tail.filterNot(g => tree.contains(_.groupId == g.groupId)), tree :: done)
+          )
+      }
+    }
+
+    for {
+      all   <- targets.accumulatePure(_.toGroupProp.map(g => (g.groupId, g))).map(_.toMap)
+      trees <- recurseTrees(all, all.values.toList, Nil)
+    } yield {
+      trees
+    }
+  }
+
+
+  /*
+   * Build a tree from given group with the `allTargets` context which is supposed to
+   * contain all group related to that one and more.
+   * Algorithm to build hierarchy:
+   * - take one of the group at random (the one given in parameter)
+   * - search for root, ie recursively search for one child having parameter group as parent
+   *     - 0 means we have the last child: root
+   *     - 1 means more children! Recurse.
+   *     - n > 1 means a diamond hierarchy. Take one branche at random, the other
+   *       one will be processed in another call to parent method. Recurse!
+   *
+   * When we are at root, we need to go back up, actually building the tree with all
+   * its branches and not only the one we took during descent:
+   *   - look at each line in criterion (we only have subgroup criterion here), keeping line
+   *     order (because it defines override order):
+   *     - 0 means we have the outer most parent
+   *     - n > 0 means: more parents!
+   */
+  def buildOneTree(group: GroupProp, allTargets: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
+    /*
+     * Descend to root
+     */
+    @scala.annotation.tailrec
+    def searchDown(current: GroupProp, all: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
+      all.collect { case g if g._2.parentGroups.contains(current.groupId.value) => g._2 } match {
+        case Nil      => // found root, go up now
+          buildUp(current, all)
+        case h :: _ => // recurse
+          searchDown(h, all)
+      }
+    }
+
+    /*
+     * Here, we look up for parent groups. We know that "all" contains all parents.
+     */
+    def buildUp(root: GroupProp, all: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
+      // recursively takes parent, keeping criterion line order
+      for {
+        parents <- root.parentGroups.accumulatePure { id =>
```

Comment:
What happens if there is a cycle between two groups ?

It should not happen, because in this case there is lot of chance that that the groups are empty (if combined with an AND),

or the composition is OR and then the groups are filtered from parents and it will trigger an error in case of common property...

So it should be ok

Reply:
cycle are forbidden by construction and check when a group is modified